assemble.go

```diff
@@ -118,7 +135,30 @@ func AssembleFile(ctx context.Context, name string, idx Index, s Store, seeds []
 					ss.add(job.segment)
 					continue
 				}
+
+				// If we don't have a seedSegment we expect an IndexSegment with just
+				// a single chunk, that we can take from either the selfSeed, from the
+				// destination file, or from the store.
+				if len(job.segment.chunks()) != 1 {
+					panic("Received an unexpected segment that doesn't contain just a single chunk")
+				}
 				c := job.segment.chunks()[0]
+
+				// If we already took this chunk from the store we can reuse it by looking
+				// into the selfSeed.
+				if segment := ss.getChunk(c.ID); segment != nil {
+					copied, cloned, err := segment.WriteInto(f, c.Start, c.Size, blocksize, isBlank)
+					if err != nil {
+						return err
+					}
+					stats.addBytesCopied(copied)
+					stats.addBytesCloned(cloned)
+					// This is just a single chunk, and we already confirmed it to be
+					// present in the self-seed. For this reason there is no need to
+					// record this segment again as being written in the self-seed.
+					continue
```

Comment:
Sorry, I may have been wrong about not needing to "add" to the selfseed here. Take a look at https://github.com/folbricht/desync/blob/master/selfseed.go#L45-L65 So basically, the selfseed previously supported matching on multi-chunk segments, similar to the other seeds. Now, to achieve that it kept a pointer into the index describing up to which point the selfseed is fully populated. Only the segments below this pointer could be used for seeding. So now if we don't "add" all the chunks to it, the pointer won't advance past a certain stage and the seed will be less useful. We could:
1) Keep that code in selfseed and put the "add" back here
2) Or just remove that logic from the self-seed and have a different way of remembering which chunks are available in the selfseed.

I suspect that 1) may be the easiest option at this point, but this could be revisited in another PR. Thoughts?
(Sorry for not realizing that issue earlier)

Reply:
Ops, you suggestion made a lot of sense and I implemented it without double checking the side effects. Sorry about that.

I implemented 1), adding back the `ss.add()` and explaining why we need that.
I prefer not to add too much changes in a single commit. With a followup MR we can always refactor how the self-seed is generated and maybe optimize it a little more by avoiding the need to add chunks that are already in the self-seed.